distributedcombigrid/src/sgpp/distributedcombigrid/fullgrid/DistributedFullGrid.hpp

```diff
@@ -201,12 +201,103 @@ class DistributedFullGrid {
+      // get distance between coords and point
+      pointCoords[d] -= coords[d];
+      if (std::abs(pointCoords[d]) > h[d]){
+        std::cout << "assert bounds " << pointCoords << coords <<
+         h << d << localIndex << lastIndex << std::endl;
+        assert(false &&
+          "should only be called for coordinates within the support of this point's basis function");
+      }
+      phi_c *= 1. - std::abs(pointCoords[d]/h[d]);
+    }
+    // std::cout << "coords " <<  localIndex << coords << localLinearIndex << h << std::endl;
+    // std::cout << "phi_c " << phi_c << this->getElementVector()[localLinearIndex] << std::endl;
+    assert(phi_c >= 0.);
+    return phi_c * this->getElementVector()[localLinearIndex];
+  }
+
+  FG_ELEMENT evalMultiindexRecursively (const IndexVector& localIndex, DimType dim, const std::vector<real>& coords) const {
+    assert(!(dim > this->getDimension()));
+    if (dim == this->getDimension()){
+      // std::cout << "eval " << localIndex << std::endl;
+      return evalLocalIndexOn(localIndex, coords);
+    } else {
+      FG_ELEMENT sum = 0.;
+      IndexVector localIndexDimPlusOne = localIndex;
+      localIndexDimPlusOne[dim] += 1;
+      // std::cout << localIndex << localIndexDimPlusOne << std::endl;
+      sum += evalMultiindexRecursively(localIndex, dim+1, coords);
+      sum += evalMultiindexRecursively(localIndexDimPlusOne, dim+1, coords);
+      return sum;
+    }
+  }
+
   /** evaluates the full grid on the specified coordinates
    * @param coords ND coordinates on the unit square [0,1]^D*/
-  FG_ELEMENT eval(std::vector<double>& coords) const {
-    assert(!"not implemented");
+  FG_ELEMENT eval(const std::vector<real>& coords) const {
+    FG_ELEMENT value;
+    eval(coords, value);
+    return value;
+  }
+
+  void eval(const std::vector<real>& coords, FG_ELEMENT& value, MPI_Request* request = nullptr) const {
+    assert(coords.size() == this->getDimension());
+
+    // get the lowest-index point of the points
+    // whose basis functions contribute to the interpolated value
+    auto lowerCoords = getLowerBoundsCoords();
+    auto h = getGridSpacing();
+    IndexVector localIndexLowerNonzeroNeighborPoint (dim_);
+    for (DimType d = 0 ; d < dim_ ; ++d){
+      assert(coords[d] >= 0. && coords[d] <= 1.);
+      localIndexLowerNonzeroNeighborPoint[d] = std::floor((coords[d] - lowerCoords[d]) / h[d]);
+    }
+    // std::cout <<localIndexLowerNonzeroNeighborPoint << coords << lowerCoords << h << std::endl;
+
+    // evaluate at those points and sum up according to the basis function
+    // needs to be recursive in order to be dimensionally adaptive
+    value = evalMultiindexRecursively(localIndexLowerNonzeroNeighborPoint, 0, coords);
 
-    return FG_ELEMENT(0);
+    if (request == nullptr) {
+      MPI_Allreduce(MPI_IN_PLACE, &value, 1, this->getMPIDatatype(), MPI_SUM, this->getCommunicator());
+    } else {
+      //TODO (pollinta) we could also MPI_Reduce only to the master rank
```

Comment:
Be aware that if you do this then also make the copying of the results in Processgroupworker Master exlcusive. Otherwise nonsense is combined there which could lead to bugs in the future.

Reply:
I don't get the comment yet -- "this" is the IAllreduce? or the TODO ?

Reply:
It is for the TODO. The return of the method is used to combine the interpolation in ProcessGroupWorker. If this is only reduced on the master this combine would use false values on the others which could cause bugs in the future as it is not obvious that the method only returns the right solutions for the master.

Reply:
valid point. I removed that TODO.